Use NetworkConnector directly to account for replaced hosts

nox opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16131)
<!-- Reviewable:end -->

nox committed:
Rework replace_hosts and host_replacement

They do less cloning now.

nox committed:
Remove some useless Option<T> wrappers from ServoUrl methods

nox committed:
Use NetworkConnector directly to account for replaced hosts

This let us remove a hack where we had to replace the host in the request URL.

highfive commented:
Heads up! This PR modifies the following files:
 * @fitzgen: components/script/dom/url.rs, components/script/dom/htmlformelement.rs, components/script/dom/urlhelper.rs
 * @kichjang: components/net/fetch/methods.rs, components/net/http_loader.rs, components/script/dom/url.rs, components/net/websocket_loader.rs, components/script/dom/htmlformelement.rs, components/script/dom/urlhelper.rs, components/net_traits/hosts.rs, components/net_traits/hosts.rs

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 8df56690b7a0cd3c991f5f76751bc5cb6ceccdea with merge cd6bc6fcaa689bfdc06d52c9e8412ebd499064d5...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3013)

nox committed:
Use Url::port instead of Url::port_or_known_default to set Host

AFAIK, if there is no explicit port in the request URL, there should be
no explicit port in the Host header.

nox commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit af19004e9febab645622653f534761dff0c93c07 with merge 152d6ce17f76e16275b8a0bf13703b85dbe61490...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3075)

nox committed:
Put websocket_loader::init first in its module

nox commented:
The failure is #14323.

nox commented:
r? @jdm

jdm commented on components/url/lib.rs:
```diff
@@ -49,8 +49,8 @@ impl ServoUrl {
 
     // NOTE: These methods return options that are always true temporarily until
     // we special-case some urls to avoid going through rust-url.
```
This comment should go away if we're making this change. I don't understand it though; any comments, @emilio?

nox replied:
@jdm This was from a time where we thought `ServoUrl` wouldn't always contain an `Url`, will remove it.

jdm reviewed (dismissed):
This is a good change. I only have one comment about the changes in 22e4c9c.

nox commented:
@jdm r=you?

jdm commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 8796a82 has been approved by `jdm`

<!-- @bors-servo r=jdm 8796a82b9fe1654c4e022110cee024e411aa3ee9 -->

bors-servo commented:
:hourglass: Testing commit 8796a82b9fe1654c4e022110cee024e411aa3ee9 with merge 1154bf56df5238e599d538fd50c2e4eebc45d39d...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3049)

nox commented:
@bors-servo retry #14237

bors-servo commented:
:hourglass: Testing commit 8796a82b9fe1654c4e022110cee024e411aa3ee9 with merge 1105100c3beb9cff3a71677937269ecc523f2924...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5979), [arm32](http://build.servo.org/builders/arm32/builds/6200), [arm64](http://build.servo.org/builders/arm64/builds/6233), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6335), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/3054), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/3083), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6339), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6368), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/3022), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/3090), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1507)
Approved by: jdm
Pushing 1105100c3beb9cff3a71677937269ecc523f2924 to master...
